### PR TITLE
[RfR] Make ListFilterMixin work

### PR DESCRIPTION
**Pull request by samchrisinger**
# Purpose

Using set intersection in ListFilterMixin was not a reliable way to filter lists. Also the sort function in ODMOrderingFillter did not work for non-querysets

# Changes

- use list comp instead of set intersection
- let ODMOrderingFilter filters lists

# Side Effects

None

### CODE REVIEW

File: api/base/filters.py [hunk 1]
```diff
@@ -171,7 +188,11 @@ def param_queryset(self, query_params, default_queryset):
         if fields_dict:
             for field_name, value in fields_dict.items():
                 if self.is_filterable_field(key=field_name):
-                    queryset = queryset.intersection(set(self.get_filtered_queryset(field_name, value, default_queryset)))
+                    filtered_queryset = set(self.get_filtered_queryset(field_name, value, default_queryset))
+                    queryset = [
+                        result for result in queryset
+                        if result in filtered_queryset
+                    ]
```
Comment (cwisecarver) [inline api/base/filters.py]: I don't think I'm understanding the value of 192-195. The `default_queryset` is passed in, `queryset = set(default_queryset)`. Then we pass `default queryset` to `self.get_filtered_queryset` and wrap the return in a `set` and assign it to `filtered_queryset`. All well and good, now we have a set of values from `default_queryset` filtered by the requested params. Then we loop through `queryset` which is a set of `default_queryset`, which we passed to `get_filtered_queryset` and see if each value is in the filtered version and reset `queryset` to that combination in a list comprehension. I know it was there before in a different form. I'm just not sure what the value of either of them is.

Comment (samchrisinger) [inline api/base/filters.py]: Agreed this code is confusing. And I may have accidentally inverted the meaning. 

Trying to think this through:

1. `queryset` needs to be the intersection of each of the `filtered_queryset`s, and starts out as the unfiltered `default_queryset`. 
2. Each `filtered_queryset` is a subset of `default_queryset`. So at each iteration of the loop we want to keep all of the members of `queryset` that are also members of `filtered_queryset`
```python
[result for result in queryset if result in filtered_queryset]
```
3. `queryset` is iteratively reduced by the constraint above

Does this help at all?

File: api/base/filters.py [hunk 2]
```diff
@@ -3,12 +3,26 @@
...
+        while fields:
+            field = fields.pop(0)
+            a_field = getattr(a, field)
+            b_field = getattr(b, field)
+            if a_field > b_field:
+                return 1
+            elif a_field < b_field:
+                return -1
+        return 0
+    return sort_fn
```
Comment (cwisecarver) [inline api/base/filters.py]: Writing your own cmp function is very slow according to the python 2.7 docs. It says each element will be touched twice with a cmp function while standard sorted and sorted with a key function only touches each once. Using a key function on the other hand is very fast. There is an attrgetter function built into the operator library. 

`return sorted(queryset, key=operator.attrgetter(*ordering))` seems to kill about 12 LOC with the same functionality and according to the docs more speed.

Comment (samchrisinger) [inline api/base/filters.py]: Awesome thanks! Wish I'd known about attrgetter long ago.

File: api/base/filters.py (hunk 1 above)
Comment (cwisecarver) [inline api/base/filters.py]: Great explanation. I totally missed the existence of line 190.

I feel like there's got to be a way where we can loop through the result set fewer times but nothing is jumping out at me right now. This works and does what it should. I'd say it's potentially ripe for optimization though. Anyone filtering by more than one field with a result set > 0 is going to be sitting in this function for a while.

Two more tiny things. 
- query_params_to_fields returns a dictionary unless an exception occurs so you could kill 189 and just let 190 loop zero times if the dict is empty.
- @abought pointed out that the end of 211 could potentially call `.lower()` on `None` which wouldn't work out well for anyone. I know that's not part of your changes but it's probably worth replacing `None` with `''` or something equally benign while you're in there.

**Comment by cwisecarver** [comment]
I'm happy. Moving to CR2.

**Comment by brianjgeiger** [comment]
By and large this looks good, though I feel like, since we're fixing bugs, we should probably have unit tests that reflect those bug fixes. Can that happen? Pass Complete. :penguin:

**Pull request closed without merging**